base/system/winlogon/sas.c

```diff
@@ -279,153 +279,305 @@ PlaySoundRoutine(
+
+    if (lRegStatus != ERROR_SUCCESS)
     {
-        /* No sound has been set */
-        ERR("No sound has been set\n");
         return 0;
     }
 
-    /* Expand environment variables */
-    if (!ExpandEnvironmentStringsW(wszBuffer, wszDest, MAX_PATH))
+    /* Open .Current */
+    lRegStatus = RegOpenKeyExW(hRegKey, L".Current", 0, KEY_QUERY_VALUE, &hRegSnd);
+
+    if (lRegStatus != ERROR_SUCCESS)
     {
-        ERR("ExpandEnvironmentStringsW failed!\n");
+        /* If fail then open .Default */
+        lRegStatus = RegOpenKeyExW(hRegKey, L".Default", 0, KEY_QUERY_VALUE, &hRegSnd);
+
+        RegCloseKey(hRegKey);
+
+        if (lRegStatus != ERROR_SUCCESS)
+        {
+            return 0;
+        }
+    }
+
+    RegCloseKey(hRegKey);
+
+    /* Get registry value size */
+    lRegStatus = RegQueryValueExW(hRegSnd, NULL, NULL, &dwRegValType, NULL, &dwRegValSize);
+
+    /* Check whether the type is valid */
+    if (lRegStatus != ERROR_SUCCESS || (dwRegValType != REG_SZ && dwRegValType != REG_EXPAND_SZ))
+    {
+        RegCloseKey(hRegSnd);
         return 0;
     }
 
-    /* Open the service manager */
-    hSCManager = OpenSCManager(NULL, NULL, SC_MANAGER_CONNECT);
-    if (!hSCManager)
+    /* Allocate buffer for registry value */
+    lpRegVal = HeapAlloc(GetProcessHeap(), 0, (SIZE_T)dwRegValSize);
+
+    if (!lpRegVal)
     {
-        ERR("OpenSCManager failed (%x)\n", GetLastError());
+        RegCloseKey(hRegSnd);
         return 0;
     }
 
-    /* Open the wdmaud service */
-    hService = OpenServiceW(hSCManager, L"wdmaud", GENERIC_READ);
-    if (!hService)
+    /* Get registry value */
+    RegQueryValueExW(hRegSnd, NULL, NULL, &dwRegValType, (LPBYTE)lpRegVal, &dwRegValSize);
+
+    RegCloseKey(hRegSnd);
+
+    /* Get full sound path length (including the terminating null character) */
+    dwSndPathLen = ExpandEnvironmentStringsW(lpRegVal, NULL, 0);
+
+    if (dwSndPathLen == 0)
     {
-        /* The service is not installed */
-        TRACE("Failed to open wdmaud service (%x)\n", GetLastError());
-        CloseServiceHandle(hSCManager);
+        HeapFree(GetProcessHeap(), 0, lpRegVal);
         return 0;
     }
 
-    /* Wait for wdmaud to start */
-    do
+    /* Allocate buffer for sound path */
+    lpSndPath = HeapAlloc(GetProcessHeap(), 0, (SIZE_T)(dwSndPathLen * sizeof(WCHAR)));
+
+    if (!lpSndPath)
     {
-        if (!QueryServiceStatusEx(hService, SC_STATUS_PROCESS_INFO, (LPBYTE)&Info, sizeof(SERVICE_STATUS_PROCESS), &dwSize))
+        HeapFree(GetProcessHeap(), 0, lpRegVal);
+        return 0;
+    }
+
+    ExpandEnvironmentStringsW(lpRegVal, lpSndPath, dwSndPathLen);
+
+    TRACE("Playing logon sound: %ls\n", lpSndPath);
+
+    /* Play sound */
+    PlaySoundRoutine(lpSndPath, TRUE, SND_FILENAME | SND_NODEFAULT);
+
+    HeapFree(GetProcessHeap(), 0, lpRegVal);
+    HeapFree(GetProcessHeap(), 0, lpSndPath);
+
+    return 0;
+}
+
+DWORD WINAPI PlayLogoffSoundThread(LPVOID lpParameter)
```

Comment:
Both logon and logoff functions seem to be quite similar; isn't it possible to merge both together, and have the actual different behaviour be controlled by a parameter?

Reply:
Updated.